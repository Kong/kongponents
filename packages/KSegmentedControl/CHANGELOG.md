# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.4](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@1.0.3...@kongponents/ksegmentedcontrol@1.0.4) (2020-10-22)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## [1.0.3](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@1.0.2...@kongponents/ksegmentedcontrol@1.0.3) (2020-10-19)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## [1.0.2](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@1.0.1...@kongponents/ksegmentedcontrol@1.0.2) (2020-07-13)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## [1.0.1](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@1.0.0...@kongponents/ksegmentedcontrol@1.0.1) (2020-06-30)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





# [1.0.0](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@0.1.4...@kongponents/ksegmentedcontrol@1.0.0) (2020-06-25)


### Features

* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [0.1.4](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@0.1.3...@kongponents/ksegmentedcontrol@0.1.4) (2020-06-08)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## [0.1.3](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@0.1.2...@kongponents/ksegmentedcontrol@0.1.3) (2020-05-20)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## [0.1.2](https://github.com/Kong/kongponents/compare/@kongponents/ksegmentedcontrol@0.1.1...@kongponents/ksegmentedcontrol@0.1.2) (2020-05-12)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol





## 0.1.1 (2020-05-01)

**Note:** Version bump only for package @kongponents/ksegmentedcontrol
