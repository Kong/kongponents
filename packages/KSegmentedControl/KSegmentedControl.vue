<template>
  <div class="segmented-control d-flex" >
    <KButton
      v-for="item in options"
      :key="label(item)"
      :name="value(item)"
      :disabled="disabled(item)"
      :appearance="appearance(item)"
      size="small"
      class="justify-content-center"
      v-on="listeners"
    >
      {{ label(item) }}
    </KButton>
  </div>
</template>

<script>
import KButton from '@kongponents/kbutton/KButton.vue'

export default {
  name: 'KSegmentedControl',
  components: { KButton },
  model: {
    prop: 'selected',
    event: 'click'
  },
  props: {
    options: {
      type: Array,
      required: true
    },
    selected: {
      type: [String, Number, Boolean],
      required: true
    },
    isDisabled: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      selectedValue: this.selected
    }
  },
  computed: {
    listeners () {
      return {
        ...this.$listeners,
        click: e => {
          this.$emit('click', e.target.name)
        }
      }
    }
  },
  methods: {
    label (item) {
      return item.label || item
    },
    value (item) {
      return item.value || item
    },
    appearance (item) {
      return this.selected === this.value(item) ? 'primary' : 'outline'
    },
    disabled (item) {
      return item.disabled || this.isDisabled
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/variables';

.segmented-control .k-button {
  --KButtonPrimaryBase: var(--blue-100);
  --KButtonPrimaryHover: var(--blue-100);

  color: var(--KSegementedControlPrimary, var(--blue-500, color(blue-500)));
  border-radius: 0;
  margin-left: -1px;
  flex: 1;

  &.primary {
    z-index: 1;
    border-color: var(--blue-500);
  }
  &:hover {
    z-index: 2;
  }
  &:active {
    z-index: 2;
  }
  &:focus {
    z-index: 3;
    box-shadow: 0 0 0 2px var(--white), 0 0 0 4px var(--blue-500);
  }
  &:first-child {
    border-radius: 3px 0 0 3px;
    margin-left: 0;
  }
  &:last-child {
    border-radius: 0 3px 3px 0;
  }
  &:only-child {
    border-radius: 3px;
    margin-left: 0;
  }
  &:disabled {
    border-color: var(--grey-500);
  }
}
</style>
