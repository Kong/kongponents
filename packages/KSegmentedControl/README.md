# @kongponents/ksegmentedcontrol

[![](https://img.shields.io/npm/v/@kongponents/ksegmentedcontrol.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/ksegmentedcontrol)

```html
<template>
<KSegmentedControl
    :options="['one','two']"
    :v-model="selected"
    @click="handleToggle"
/>
</template>

<script>
export default {
  data() {
    return {
      selected: 'one'
    }
  },
  methods: {
    handleToggle(selected) {
      this.selected = selected
      // do something, make api call?
    }
  }
}
</script>
```
