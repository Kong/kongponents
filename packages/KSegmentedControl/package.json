{
  "name": "@kongponents/ksegmentedcontrol",
  "version": "6.26.3",
  "description": "Segmented Control component.",
  "main": "dist/KSegmentedControl.umd.min.js",
  "componentName": "KSegmentedControl",
  "source": "KSegmentedControl.vue",
  "files": [
    "dist",
    "KSegmentedControl.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^6.26.3",
    "@kongponents/styles": "^6.26.3"
  }
}
