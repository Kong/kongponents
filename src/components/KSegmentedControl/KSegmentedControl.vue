<template>
  <div class="k-segmented-control">
    <button
      v-for="option in normalizedOptions"
      :key="`${option.value}-option`"
      class="segmented-control-button"
      :class="[size, { selected: props.modelValue === option.value }]"
      :data-testid="`${option.value}-option`"
      :disabled="getDisabled(option)"
      type="button"
      @click="handleClick(option)"
    >
      <slot
        name="option-label"
        :option="option"
      >
        {{ option.label }}
      </slot>
    </button>
  </div>
</template>

<script lang="ts">
import type { PropType } from 'vue'
import { ref } from 'vue'
import type { SegmentedControlOption } from '@/types/segmented-control'

const itemsHaveRequiredProps = (items: SegmentedControlOption[]): boolean => {
  return items.every(i => i.value !== undefined)
}

// functions used in prop validators
const getValues = (items: SegmentedControlOption[]) => {
  const vals: string[] = []
  items.forEach((item: SegmentedControlOption) => vals.push(item.value + ''))

  return vals
}

const itemValuesAreUnique = (items: SegmentedControlOption[]): boolean => {
  const vals = getValues(items)
  const uniqueValues = new Set(vals)

  return vals.length === uniqueValues.size
}

const normalizeItems = (items: SegmentedControlOption[] | string[]): SegmentedControlOption[] => {
  return items.map((item:SegmentedControlOption | string) => {
    return {
      label: typeof item === 'string' ? item : (item.label || (item.value + '')),
      value: typeof item === 'string' ? item.toLocaleLowerCase().replace(' ', '-') : item.value,
      disabled: typeof item === 'string' ? false : item.disabled,
    } as SegmentedControlOption
  })
}

const validateItems = (items: SegmentedControlOption[] | string[]): boolean => {
  const isStringArray = typeof items[0] === 'string'
  const nItems = normalizeItems(items)
  const isValid = itemValuesAreUnique(nItems)

  return isStringArray ? isValid && itemsHaveRequiredProps(nItems as SegmentedControlOption[]) : isValid
}

export default {}
</script>

<script lang="ts" setup>
const props = defineProps({
  modelValue: {
    type: [String, Number, Boolean],
    required: true,
  },
  options: {
    type: Array as PropType<SegmentedControlOption[] | string[]>,
    required: true,
    validator: (items: SegmentedControlOption[] | string[]) => !items.length || validateItems(items),
  },
  size: {
    type: String,
    default: 'small',
    validator: (size: string) => ['small', 'large'].includes(size),
  },
  disabled: {
    type: Boolean,
    default: false,
  },
})

const emit = defineEmits<{
  (e: 'click', event: string): void;
  (e: 'update:modelValue', event: string): void;
}>()

const normalizedOptions = ref(normalizeItems(props.options))

const getDisabled = (option: SegmentedControlOption): boolean => {
  return !!option.disabled || props.disabled
}

const handleClick = (option: SegmentedControlOption): void => {
  // @ts-ignore: allow emitting a custom value
  emit('click', option.value)
  // @ts-ignore: allow emitting a custom value
  emit('update:modelValue', option.value)
}
</script>

<style lang="scss" scoped>
/* Component variables */

$kSegmentedControlSmallHeight: 32px;

/* Component styles */

.k-segmented-control {
  display: flex;
  gap: var(--kui-space-0, $kui-space-0);
  width: 100%;

  .segmented-control-button {
    align-items: center;
    background-color: var(--kui-color-background, $kui-color-background);
    border-color: var(--kui-color-border-primary, $kui-color-border-primary);
    border-style: solid;
    border-width: var(--kui-border-width-20, $kui-border-width-20);
    color: var(--kui-color-text-primary, $kui-color-text-primary);
    cursor: pointer;
    display: flex;
    font-family: var(--kui-font-family-text, $kui-font-family-text);
    font-size: var(--kui-font-size-20, $kui-font-size-20);
    font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
    gap: var(--kui-space-30, $kui-space-30);
    height: $kSegmentedControlSmallHeight;
    justify-content: center;
    line-height: var(--kui-line-height-20, $kui-line-height-20);
    outline: none;
    padding-left: var(--kui-space-50, $kui-space-50);
    padding-right: var(--kui-space-50, $kui-space-50);
    transition: border-color $kongponentsTransitionDurTimingFunc, color $kongponentsTransitionDurTimingFunc, background-color $kongponentsTransitionDurTimingFunc, box-shadow $kongponentsTransitionDurTimingFunc;
    white-space: nowrap;
    width: 100%;
    z-index: 1;

    &:not(:first-child) {
      // offset the border of the previous button
      /* stylelint-disable-next-line @kong/design-tokens/use-proper-token */
      margin-left: calc(var(--kui-border-width-20, $kui-border-width-20) * -1);
    }

    &:first-child {
      border-bottom-left-radius: var(--kui-border-radius-30, $kui-border-radius-30);
      border-top-left-radius: var(--kui-border-radius-30, $kui-border-radius-30);
    }

    &:last-child {
      border-bottom-right-radius: var(--kui-border-radius-30, $kui-border-radius-30);
      border-top-right-radius: var(--kui-border-radius-30, $kui-border-radius-30);
    }

    &.large {
      height: $kongponentsInputElementHeight;
      padding-left: var(--kui-space-60, $kui-space-60);
      padding-right: var(--kui-space-60, $kui-space-60);
    }

    &:hover:not([disabled]) {
      border-color: var(--kui-color-border-primary-strong, $kui-color-border-primary-strong);
      color: var(--kui-color-text-primary-strong, $kui-color-text-primary-strong);
      z-index: 2;
    }

    &:focus:not([disabled]) {
      border-color: var(--kui-color-border-primary-stronger, $kui-color-border-primary-stronger);
      color: var(--kui-color-text-primary-stronger, $kui-color-text-primary-stronger);
      z-index: 3;
    }

    &:active:not([disabled]) {
      border-color: var(--kui-color-border-primary-strongest, $kui-color-border-primary-strongest);
      color: var(--kui-color-text-primary-strongest, $kui-color-text-primary-strongest);
      z-index: 3;
    }

    &:focus-visible:not([disabled]) {
      border-color: var(--kui-color-border-primary-strong, $kui-color-border-primary-strong);
      box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
      color: var(--kui-color-text-primary-strong, $kui-color-text-primary-strong);
      z-index: 3;
    }

    &[disabled] {
      border-color: var(--kui-color-border-disabled, $kui-color-border-disabled) !important;
      color: var(--kui-color-text-disabled, $kui-color-text-disabled) !important;
      cursor: not-allowed;
      z-index: 0;

      &.selected {
        background-color: var(--kui-color-background-disabled, $kui-color-background-disabled);
      }
    }

    &.selected {
      background-color: var(--kui-color-background-primary-weakest, $kui-color-background-primary-weakest);
      border-color: var(--kui-color-border-primary-stronger, $kui-color-border-primary-stronger);
      color: var(--kui-color-text-primary-stronger, $kui-color-text-primary-stronger);
      z-index: 2;
    }

    :deep(#{$kongponentsKongIconSelector}) {
      height: var(--kui-icon-size-40, $kui-icon-size-40) !important;
      width: var(--kui-icon-size-40, $kui-icon-size-40) !important;
    }
  }
}
</style>
