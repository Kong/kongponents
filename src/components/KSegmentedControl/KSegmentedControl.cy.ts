import { mount } from 'cypress/vue'
import KSegmentedControl from '@/components/KSegmentedControl/KSegmentedControl.vue'

describe('KSegmentedControl', () => {
  it('renders custom button text & appearance', () => {
    const items = [
      { label: 'Item 1', value: 'item1' },
      { label: 'Item 2', value: 'item2' },
    ]

    mount(KSegmentedControl, {
      props: {
        options: items,
        modelValue: items[1].value,
      },
    })

    cy.get('.k-segmented-control button').eq(0).should('contain.text', items[0].label)

    cy.get('.k-segmented-control button').eq(1).should('contain.text', items[1].label)
    cy.get('.k-segmented-control button').eq(1).should('have.class', 'selected')
  })

  it('renders slotted button text & appearance', () => {
    const items = [
      { label: 'Item 1', value: 'item1' },
      { label: 'Item 2', value: 'item2' },
    ]

    mount(KSegmentedControl, {
      props: {
        options: items,
        modelValue: items[1].value,
      },
      slots: {
        'option-label': `<template #option-label="{ option }">
            Hello {{ option.label }}
            </template>
          `,
      },
    })

    cy.get('.k-segmented-control button').eq(0).should('contain.text', 'Hello ' + items[0].label)

    cy.get('.k-segmented-control button').eq(1).should('contain.text', 'Hello ' + items[1].label)
    cy.get('.k-segmented-control button').eq(1).should('have.class', 'selected')
  })

  it('disables an item if option is set to disabled', () => {
    mount(KSegmentedControl, {
      props: {
        options: [
          {
            label: 'One',
            value: 1,
          },
          {
            label: 'Two',
            value: 2,
            disabled: true,
          },
          {
            label: 'Three',
            value: 3,
          },
        ],
        modelValue: 1,
      },
    })

    cy.get('.k-segmented-control button').eq(0).invoke('attr', 'disabled').should('not.exist')
    cy.get('.k-segmented-control button').eq(1).invoke('attr', 'disabled').should('eq', 'disabled')
    cy.get('.k-segmented-control button').eq(2).invoke('attr', 'disabled').should('not.exist')
  })

  it('renders all items disabled when disabled prop is true', () => {
    mount(KSegmentedControl, {
      props: {
        options: [
          {
            label: 'One',
            value: 1,
          },
          {
            label: 'Two',
            value: 2,
          },
          {
            label: 'Three',
            value: 3,
          },
        ],
        modelValue: 1,
        disabled: true,
      },
    })

    cy.get('.k-segmented-control button').eq(0).invoke('attr', 'disabled').should('eq', 'disabled')
    cy.get('.k-segmented-control button').eq(1).invoke('attr', 'disabled').should('eq', 'disabled')
    cy.get('.k-segmented-control button').eq(2).invoke('attr', 'disabled').should('eq', 'disabled')
  })

  it('emits update:modelValue event when an item is clicked', () => {
    const items = [
      { label: 'Item 1', value: 'item1' },
      { label: 'Item 2', value: 'item2' },
    ]

    mount(KSegmentedControl, {
      props: {
        options: items,
        modelValue: items[1].value,
      },
    })

    cy.get('.k-segmented-control button').eq(0).click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'update:modelValue')
      cy.wrap(Cypress.vueWrapper.emitted('update:modelValue')?.[0][0]).should('eq', items[0].value)
    })
  })
})
