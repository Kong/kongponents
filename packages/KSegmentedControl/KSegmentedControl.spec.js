import { mount } from '@vue/test-utils'
import KSegmentedControl from '@/KSegmentedControl/KSegmentedControl'

describe('KSegmentedControl', () => {
  it('renders custom button text & appearance', () => {
    const wrapper = mount(KSegmentedControl, {
      propsData: {
        options: ["I'm a segment", 'So am I'],
        selected: 'So am I'
      }
    })
    const buttons = wrapper.find('.segmented-control').findAll('button')

    expect(buttons.at(0).text()).toBe("I'm a segment")
    expect(buttons.at(0).classes()).toContain('outline-primary')
    expect(buttons.at(1).text()).toBe('So am I')
    expect(buttons.at(1).classes()).toContain('primary')
  })

  it('matches snapshot', () => {
    const wrapper = mount(KSegmentedControl, {propsData: {options: ['1', '2']}})

    expect(wrapper.html()).toMatchSnapshot()
  })
})
