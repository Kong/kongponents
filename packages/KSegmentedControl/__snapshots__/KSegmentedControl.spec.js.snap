// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KSegmentedControl matches snapshot 1`] = `
<div class="segmented-control d-flex justify-content-between"><button type="button" class="k-button justify-content-center outline-primary">
    1
  </button><button type="button" class="k-button justify-content-center outline-primary">
    2
  </button></div>
`;
